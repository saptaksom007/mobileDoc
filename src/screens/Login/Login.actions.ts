import { Tokens } from 'api/types'

export const ActionTypes = {
  LOGIN_REQUESTED: 'LOGIN_REQUESTED',
  LOGIN_SUCCEEDED: 'LOGIN_SUCCEEDED',
  LOGIN_FAILED: 'LOGIN_FAILED',
  INITIAL_LOADING_REQUESTED: 'INITIAL_LOADING_REQUESTED',
  INITIAL_LOADING_SUCCEEDED: 'INITIAL_LOADING_SUCCEEDED',
  INITIAL_LOADING_FAILED: 'INITIAL_LOADING_FAILED',
  SESSION_HAS_EXPIRED: 'SESSION_HAS_EXPIRED',
  RELOGIN_REQUESTED: 'RELOGIN_REQUESTED',
  SET_IS_LOGIN: 'SET_IS_LOGIN',
  GET_LIMITED_ACCESS_TOKEN_REQUESTED: 'GET_LIMITED_ACCESS_TOKEN_REQUESTED',
  GET_LIMITED_ACCESS_TOKEN_SUCCEEDED: 'GET_LIMITED_ACCESS_TOKEN_SUCCEEDED',
  GET_LIMITED_ACCESS_TOKEN_FAILED: 'GET_LIMITED_ACCESS_TOKEN_FAILED',
}

export default class Actions {
  static loginRequested(url: string) {
    return {
      type: ActionTypes.LOGIN_REQUESTED,
      url,
    }
  }
  static loginSucceeded(user: Tokens) {
    return {
      type: ActionTypes.LOGIN_SUCCEEDED,
      payload: { user },
    }
  }
  static loginFailed(error: Error) {
    return {
      type: ActionTypes.LOGIN_FAILED,
      message: error.message,
      error,
    }
  }
  static initialLoadingRequested() {
    return {
      type: ActionTypes.INITIAL_LOADING_REQUESTED,
    }
  }
  static initialLoadingSucceeded() {
    return {
      type: ActionTypes.INITIAL_LOADING_SUCCEEDED,
    }
  }
  static initialLoadingFailed(error: Error) {
    return {
      type: ActionTypes.INITIAL_LOADING_FAILED,
      message: error.message,
      error,
    }
  }
  static sessionHasExpired() {
    return { type: ActionTypes.SESSION_HAS_EXPIRED }
  }
  static getLimitedAccessTokenRequested() {
    return { type: ActionTypes.GET_LIMITED_ACCESS_TOKEN_REQUESTED }
  }
  static getLimitedAccessTokenSucceeded(
    limitedAccessToken: string,
    limitedAccessTokenExp: number,
  ) {
    return {
      type: ActionTypes.GET_LIMITED_ACCESS_TOKEN_SUCCEEDED,
      limitedAccessToken,
      limitedAccessTokenExp,
    }
  }
  static getLimitedAccessTokenFailed(error: Error) {
    return {
      type: ActionTypes.GET_LIMITED_ACCESS_TOKEN_FAILED,
      message: error.message,
      error,
    }
  }
  static setIsLogin(user: Partial<Tokens> & { isLoggedIn: boolean }) {
    return {
      type: ActionTypes.SET_IS_LOGIN,
      payload: user,
    }
  }
}
