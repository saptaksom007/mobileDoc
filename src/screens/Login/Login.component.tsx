import React, { Component } from 'react'
import { WebView } from 'react-native-webview'
import { Keyboard, View, StyleSheet, StatusBar, Platform } from 'react-native'
import { connect } from 'react-redux'
import { compose } from 'recompose'
import Constants from 'expo-constants'

import { withKeyboardAvoidingViewIf } from 'screens/enhancers/withKeyboardAvoidingView'
import KeycloakUtil from 'utilities/keycloak'
import { BackButton } from 'components/BackButton/BackButton.component'
import { Loader } from 'components/Loader/Loader.component'
import { Env } from 'env'

import Actions from './Login.actions'
import styles from './Login.styles'
import { Action } from 'redux'
import { WebViewNavigation } from 'react-native-webview/lib/WebViewTypes'
import { readable } from 'common-docdok/lib/utils/crypto'
import { getRememberMeAsync } from 'api/user'

export const e2eJsToInject = () =>
  `document.getElementById("username").value = '${
    Env.e2e ? (Env.e2e as any).config.login : ''
  }';` +
  `document.getElementById("password").value = '${
    Env.e2e ? (Env.e2e as any).config.password : ''
  }';` +
  'setTimeout(function(){ document.getElementById("kc-login").click()}, 2500);'

function injectedJavaScript(rememberMeEmail?: string) {
  const removeLang = 'document.querySelector(".lang-selection").remove();'
  const removeShadow =
    'document.querySelector(".main")' +
    '.setAttribute("style", "box-shadow: none;");'
  const resizeLogo =
    'document.querySelector(".spacer + div > img")' +
    '.setAttribute("width", "133");'
  const showLogo =
    'document.querySelector(".spacer + div.logo")' +
    '.setAttribute("style", "visibility: visible");'
  const removeSpacer = 'document.querySelector(".spacer").remove();'
  const blurInput =
    'var blurPwd = ' +
    'function(){ document.getElementById("password").blur(); };'
  const onSubmit =
    'document.getElementById("kc-form-login")' +
    '.addEventListener("submit", blurPwd);'

  return (
    removeLang +
    removeShadow +
    resizeLogo +
    showLogo +
    removeSpacer.repeat(2) +
    blurInput +
    onSubmit +
    (rememberMeEmail
      ? `document.getElementById("username").value = '${rememberMeEmail}';`
      : '') +
    (!Constants.isDevice && Env.e2e ? e2eJsToInject() : '')
  )
}

interface Props {
  dispatch(action: Action): void
}

interface State {
  raceCondition: boolean
  showBackButton: boolean
  rememberMeEmail?: string | null
  loginUrl: string
}

export class Login extends Component<Props, State> {
  state = {
    raceCondition: false,
    showBackButton: false,
    rememberMeEmail: undefined,
    loginUrl: KeycloakUtil.getLoginURL(),
  }

  componentDidMount() {
    StatusBar.setHidden(true)
    this.remember()
  }

  componentWillUnmount() {
    StatusBar.setHidden(false)
  }

  onNavigationStateChange = ({ url, loading }: WebViewNavigation) => {
    const { dispatch } = this.props
    const { raceCondition } = this.state
    if (KeycloakUtil.matchAuthURL(url) && loading && !raceCondition) {
      Keyboard.dismiss()
      dispatch(Actions.loginRequested(url))
      this.setState({ raceCondition: true, showBackButton: false })
    }
  }

  onLoadEnd = (event: any) => {
    const { url, canGoBack } = event.nativeEvent
    const { loginUrl } = this.state
    this.setState({ showBackButton: url !== loginUrl && canGoBack })
  }

  async remember(): Promise<any> {
    const rememberMeEmailEncrypt = await getRememberMeAsync()
    const rememberMeEmail = readable(rememberMeEmailEncrypt!)
    this.setState({ rememberMeEmail })
  }

  WEBVIEW_REF: any = undefined

  render() {
    const {
      rememberMeEmail,
      showBackButton,
      loginUrl,
      raceCondition,
    } = this.state
    if (raceCondition) {
      return <Loader withLogo info={'Login Race Condition'} />
    }
    return (
      <View style={styles.container} testID='login'>
        <WebView
          ref={ref => (this.WEBVIEW_REF = ref)}
          source={{
            uri: loginUrl,
          }}
          renderLoading={() => (
            <Loader
              withLogo
              info={'Login WebView'}
              containerProps={{
                style: StyleSheet.absoluteFillObject,
              }}
            />
          )}
          startInLoadingState
          onNavigationStateChange={this.onNavigationStateChange}
          onLoadEnd={this.onLoadEnd}
          javaScriptEnabled
          domStorageEnabled
          injectedJavaScript={injectedJavaScript(rememberMeEmail)}
          mixedContentMode={'always'}
          sharedCookiesEnabled
        />
        <BackButton
          onPress={() => {
            this.WEBVIEW_REF.goBack()
            this.setState({ showBackButton: false })
          }}
          visible={showBackButton}
        />
      </View>
    )
  }
}

function select({ login: { error } }: any) {
  return { error }
}

export default compose<Props, any>(
  withKeyboardAvoidingViewIf(Platform.OS === 'android'),
  connect(select),
)(Login)
