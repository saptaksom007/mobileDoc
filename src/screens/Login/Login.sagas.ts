import { call, put, takeEvery, select, all, delay } from 'redux-saga/effects'
import { pathOr, pipe, path, map, prop } from 'ramda'
import { Actions as NavAction } from 'navigation/SagaNavigation'
import i18n from 'ex-react-native-i18n'
import * as Sentry from 'sentry-expo'
import * as Updates from 'expo-updates'
import { messagingActions } from 'common-docdok/lib/domain/messaging/actions'
import {
  healthrelationActions,
  healthrelationActionTypes,
} from 'common-docdok/lib/domain/healthrelation/actions'
import {
  userActions,
  userActionTypes,
} from 'common-docdok/lib/domain/user/actions'
import { surveyActions } from 'common-docdok/lib/domain/survey/actions'
import { PersistStoreActions } from 'common-docdok/lib/configuration/persistStore'
import { asyncAction } from 'common-docdok/lib/utils/sagaUtils'
import { KeycloakInfo } from 'utilities/keycloak'
import Actions, { ActionTypes } from './Login.actions'
import DashboardActionsfrom from 'screens/Dashboard/Dashboard.actions'
import AppointmentsListActions from 'screens/AppointmentsList/AppointmentsList.actions'
import { LIMITED_ACCESS_TOKEN_KEY, TOKEN_KEY } from 'config/auth'
import {
  deleteStorageItem,
  setStorageItemEncrypted,
} from 'utilities/localStorage'
import {
  retrieveTokensAsync,
  persistExpireInAsync,
  getLimitedAccessTokenFromStoreAsync,
  getLimitedAccessTokenExpirationFromStoreAsync,
  getLimitedAccessTokenAsync,
} from 'api/auth'
import { Action } from 'redux'

const isLoggedInSelector = pathOr(false, ['login', 'isLoggedIn'])
const getNextRoute = pathOr(undefined, ['dashboard', 'nextRoute'])

function* gotoNextRoute() {
  try {
    const nextRoute: Action = yield select(getNextRoute)
    if (nextRoute) {
      yield put(nextRoute)
      yield put(DashboardActionsfrom.nextRouteSucceeded())
    } else {
      yield put(NavAction.navigate('App'))
    }
  } catch (error) {
    yield put(DashboardActionsfrom.nextRouteFailed(error))
  }
}

function* retrieveTokens({ url }: any) {
  try {
    const {
      access_token,
      refresh_token,
      expires_in,
      refresh_expires_in,
    } = yield call(retrieveTokensAsync, url)

    if (
      typeof access_token !== 'undefined' &&
      typeof refresh_token !== 'undefined'
    ) {
      const user = {
        access_token,
        refresh_token,
        expires_in,
        refresh_expires_in,
      }

      KeycloakInfo.token = access_token
      KeycloakInfo.refreshToken = refresh_token
      KeycloakInfo.expiresIn = expires_in

      yield call(persistExpireInAsync, expires_in)
      yield call(setStorageItemEncrypted, TOKEN_KEY, JSON.stringify(user))
      yield put(Actions.loginSucceeded(user))
      yield asyncAction(Actions.getLimitedAccessTokenRequested())
      yield asyncAction(Actions.initialLoadingRequested())
      yield call(gotoNextRoute)
    } else {
      if (!__DEV__) {
        Sentry.Native.captureException(
          new Error('access_token or refresh_token undefined'),
        )
      }
      yield put(NavAction.immediatelyResetStack(['error']))
    }
  } catch (e) {
    yield put(Actions.loginFailed(e))
    yield put(
      NavAction.showLocalError(i18n.t('login.errors.retrieveTokens'), 'login'),
    )
  }
}

function* loginError() {
  try {
    yield put(NavAction.immediatelyResetStack(['error']))
  } catch (e) {
    yield put(Actions.loginFailed(e))
  }
}

function* getLimitedAccessTokenFromStorage() {
  try {
    const limitedAccessToken: string = yield call(
      getLimitedAccessTokenFromStoreAsync,
      LIMITED_ACCESS_TOKEN_KEY,
    )
    const limitedAccessTokenExp: number = yield call(
      getLimitedAccessTokenExpirationFromStoreAsync,
    )

    if (
      limitedAccessToken &&
      limitedAccessTokenExp &&
      Date.now() < Number(limitedAccessTokenExp)
    ) {
      yield put(
        Actions.getLimitedAccessTokenSucceeded(
          limitedAccessToken,
          limitedAccessTokenExp,
        ),
      )
    } else {
      const isLogin: boolean = yield select(isLoggedInSelector)
      if (!isLogin) {
        yield put(AppointmentsListActions.appointmentslistLoaded())
        yield put(NavAction.push('Auth'))
      }
    }
  } catch (e) {
    yield put(Actions.getLimitedAccessTokenFailed(e))
    yield put(
      NavAction.showLocalError(
        i18n.t('login.errors.getLimitedAccessTokenAsync'),
      ),
    )
  }
}

const getLimitedAccessToken = function* getLimitedAccessTokenFn() {
  try {
    const isLogin: boolean = yield select(isLoggedInSelector)
    if (isLogin) {
      const { limitedAccessToken, limitedAccessTokenExp } = yield call(
        getLimitedAccessTokenAsync,
      )
      yield put(
        Actions.getLimitedAccessTokenSucceeded(
          limitedAccessToken,
          limitedAccessTokenExp,
        ),
      )
    } else {
      yield call(getLimitedAccessTokenFromStorage)
    }
  } catch (e) {
    yield put(Actions.getLimitedAccessTokenFailed(e))
    yield put(
      NavAction.showLocalError(
        i18n.t('login.errors.getLimitedAccessTokenAsync'),
      ),
    )
  }
}

const initialLoading = function* initialLoadingFn() {
  try {
    yield asyncAction(userActions.profileFetchRequested())
    yield asyncAction(messagingActions.getSubscriptionsRequested())
    yield asyncAction(healthrelationActions.loadAllPatientsRequested())
    yield asyncAction(healthrelationActions.getAvailabilitiesRequested())
    yield asyncAction(surveyActions.getMySurveysRequested())
    yield put(Actions.initialLoadingSucceeded())
  } catch (e) {
    yield put(Actions.initialLoadingFailed(e))
    yield put(NavAction.showLocalError(i18n.t('login.errors.getMe')))
  }
}

const relogin = function* reloginFn() {
  yield put(PersistStoreActions.stopPersistStore())
  yield delay(1500)
  yield call(Updates.reloadAsync)
}

function initSentry({ payload: { id } }: any) {
  if (!__DEV__) {
    Sentry.Native.setUser({ id })
  }
}

function initKeycloack({ payload: { uid } }: any) {
  KeycloakInfo.userId = uid
}

const injectStateFromPersist = function* injectStateFromPersistFn() {
  yield put(PersistStoreActions.startPersistStore())
}

const loadPatients = function* loadPatientsFn() {
  const patientIds: any[] = yield select(
    pipe(
      path<any[]>(['healthrelation', 'patients']),
      map(prop('uuid')),
    ),
  )
  const primaryPhysicianUuids: any[] = yield select(
    pipe(
      path<any[]>(['healthrelation', 'patients']),
      map(prop('primaryPhysicianUuid')),
    ),
  )
  yield all(
    patientIds.map((id: any) =>
      put(healthrelationActions.loadPatientRequested(id)),
    ),
  )
  yield put(
    healthrelationActions.loadMissingProfessionals(primaryPhysicianUuids),
  )
}

function* removeExpiresIn() {
  yield call(deleteStorageItem, 'expiresIn')
}

export default [
  takeEvery(userActionTypes.PROFILE_FETCH_SUCCEEDED, initSentry),
  takeEvery(userActionTypes.PROFILE_FETCH_SUCCEEDED, initKeycloack),
  takeEvery(
    healthrelationActionTypes.LOAD_ALL_PATIENTS_SUCCEEDED,
    loadPatients,
  ),
  takeEvery(ActionTypes.INITIAL_LOADING_REQUESTED, initialLoading),
  takeEvery(ActionTypes.INITIAL_LOADING_SUCCEEDED, injectStateFromPersist),
  takeEvery(ActionTypes.LOGIN_REQUESTED, retrieveTokens),
  takeEvery(ActionTypes.LOGIN_FAILED, loginError),
  takeEvery(ActionTypes.RELOGIN_REQUESTED, relogin),
  takeEvery(ActionTypes.GET_LIMITED_ACCESS_TOKEN_FAILED, removeExpiresIn),
  takeEvery(
    ActionTypes.GET_LIMITED_ACCESS_TOKEN_REQUESTED,
    getLimitedAccessToken,
  ),
]
