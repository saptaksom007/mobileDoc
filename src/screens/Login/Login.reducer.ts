import { ActionTypes as DashboardActionTypes } from 'screens/Dashboard/Dashboard.actions'
import { ActionTypes } from './Login.actions'

const {
  LOGIN_REQUESTED,
  LOGIN_SUCCEEDED,
  LOGIN_FAILED,
  SESSION_HAS_EXPIRED,
  RELOGIN_REQUESTED,
  GET_LIMITED_ACCESS_TOKEN_REQUESTED,
  GET_LIMITED_ACCESS_TOKEN_SUCCEEDED,
  GET_LIMITED_ACCESS_TOKEN_FAILED,
  SET_IS_LOGIN,
} = ActionTypes

const { REFRESH_TOKEN_SUCCEEDED, LOGOUT_SUCCEEDED } = DashboardActionTypes

const initialState = {
  access_token: undefined,
  refresh_token: undefined,
  expires_in: undefined,
  refresh_expires_in: undefined,
  limitedAccessToken: undefined,
  limitedAccessTokenExp: 0,
  isLimitedAccessTokenError: false,
  isLoggedIn: false,
  sessionHasExpired: false,
}

export default function login(state: any = initialState, action: any): any {
  switch (action.type) {
    case LOGIN_REQUESTED: {
      return {
        ...state,
        access_token: undefined,
        refresh_token: undefined,
        expires_in: undefined,
        refresh_expires_in: undefined,
        limited_access_token: undefined,
        isLoggedIn: false,
      }
    }

    case LOGIN_SUCCEEDED: {
      const {
        access_token,
        refresh_token,
        expires_in,
        refresh_expires_in,
      } = action.payload.user
      return {
        ...state,
        access_token,
        refresh_token,
        expires_in,
        refresh_expires_in,
        isLoggedIn: true,
        sessionHasExpired: false,
      }
    }

    case GET_LIMITED_ACCESS_TOKEN_REQUESTED: {
      return {
        ...state,
      }
    }

    case GET_LIMITED_ACCESS_TOKEN_SUCCEEDED: {
      const { limitedAccessToken, limitedAccessTokenExp } = action
      return {
        ...state,
        limitedAccessToken,
        limitedAccessTokenExp,
      }
    }

    case GET_LIMITED_ACCESS_TOKEN_FAILED: {
      return {
        ...state,
        isLimitedAccessTokenError: true,
      }
    }

    case REFRESH_TOKEN_SUCCEEDED: {
      const {
        access_token,
        refresh_token,
        expires_in,
        refresh_expires_in,
      } = action.result
      return {
        ...state,
        access_token,
        refresh_token,
        expires_in,
        refresh_expires_in,
        isLoggedIn: true,
      }
    }

    case LOGIN_FAILED: {
      return {
        ...state,
        error: action.message,
        access_token: undefined,
        refresh_token: undefined,
        expires_in: undefined,
        refresh_expires_in: undefined,
        isLoggedIn: false,
      }
    }

    case LOGOUT_SUCCEEDED: {
      const { limitedAccessToken, limitedAccessTokenExp } = state
      return {
        ...state,
        isLoggedIn: false,
        access_token: undefined,
        refresh_token: undefined,
        refresh_expires_in: undefined,
        limitedAccessToken,
        limitedAccessTokenExp,
      }
    }

    case SESSION_HAS_EXPIRED: {
      return {
        ...state,
        sessionHasExpired: true,
      }
    }

    case RELOGIN_REQUESTED: {
      return {
        ...state,
        sessionHasExpired: false,
        isLoggedIn: false,
      }
    }

    case SET_IS_LOGIN: {
      return {
        ...state,
        ...action.payload,
      }
    }

    default:
      return state
  }
}
