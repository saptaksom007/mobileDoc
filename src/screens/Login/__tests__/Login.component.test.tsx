import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { Login } from '../Login.component'

jest.mock('utilities/keycloak', () => ({
  getLoginURL() {
    return 'https://xaviercarpentier.com'
  },
}))

jest.mock('components/Loader/Loader.component', () => ({ Loader: 'Loader' }))

it('<Login/> render correctly', () => {
  const tree = renderer.create(<Login dispatch={() => {}} />).toJSON()
  expect(tree).toMatchSnapshot()
})
