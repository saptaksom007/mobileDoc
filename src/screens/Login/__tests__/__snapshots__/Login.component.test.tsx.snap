// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login/> render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="login"
>
  WebView
</View>
`;
